include/libtorrent/aux_/websocket_stream.hpp

```diff
@@ -0,0 +1,218 @@
+      from this software without specific prior written permission.
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
+LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+POSSIBILITY OF SUCH DAMAGE.
+
+*/
+
+#ifndef TORRENT_WEBSOCKET_STREAM_HPP_INCLUDED
+#define TORRENT_WEBSOCKET_STREAM_HPP_INCLUDED
+
+#include "libtorrent/config.hpp"
+
+#if TORRENT_USE_RTC
+
+#include "libtorrent/assert.hpp"
+#include "libtorrent/close_reason.hpp"
+#include "libtorrent/error_code.hpp"
+#include "libtorrent/io.hpp"
+#include "libtorrent/io_context.hpp"
+#include "libtorrent/aux_/resolver_interface.hpp"
+#include "libtorrent/ssl.hpp"
+#include "libtorrent/ssl_stream.hpp"
+#include "libtorrent/time.hpp"
+#include "libtorrent/debug.hpp"
+
+#include <boost/asio/ip/tcp.hpp>
+#include <boost/beast/core.hpp>
+#include <boost/beast/websocket.hpp>
+
+#include <functional>
+#include <vector>
+#include <string>
+#include <variant>
+
+namespace boost {
+namespace beast {
+namespace websocket {
+
+// We need to provide an overload of teardown and async_teardown
+// to use beast::websocket::stream with libtorrent::ssl::stream
+
+template<class Stream>
+void teardown(role_type
+	, libtorrent::ssl::stream<Stream>& stream
+	, error_code& ec)
+{
+    stream.shutdown(ec);
+}
+
+template<class Stream, class Handler>
+void async_teardown(role_type
+	, libtorrent::ssl::stream<Stream>& stream
+	, Handler&& handler)
+{
+    stream.async_shutdown(std::forward<Handler>(handler));
+}
+
+} // websocket
+} // beast
+} // boost
+
+namespace libtorrent {
+namespace aux {
+
+namespace websocket = boost::beast::websocket;
+
+using tcp = boost::asio::ip::tcp;
+
+struct TORRENT_EXTRA_EXPORT websocket_stream
+	: std::enable_shared_from_this<websocket_stream>
+{
+	using connect_handler = std::function<void(error_code const&)>;
+	using read_handler = std::function<void(error_code const&, std::size_t)>;
+	using write_handler = std::function<void(error_code const&, std::size_t)>;
+
+	websocket_stream(io_context& ios
+		, resolver_interface& resolver
+		, ssl::context* ssl_ctx
+	);
+
+	~websocket_stream();
+	websocket_stream& operator=(websocket_stream const&) = delete;
+	websocket_stream(websocket_stream const&) = delete;
+	websocket_stream& operator=(websocket_stream&&) noexcept = delete;
+	websocket_stream(websocket_stream&&) noexcept = delete;
```

Comment:
isn't this a member of `polymorphic_socket`? Doesn't it need to be move-constructible?

I guess not, this is only used for the tracker connection, is that right?

Reply:
You are right, it's not a member of `polymorphic_socket`, it's only used for the tracker.